ophyd/pv_positioner.py

```diff
@@ -296,3 +301,150 @@ def done_moving(**kwargs):
         else:
             self.setpoint.put(position, wait=False,
                               callback=done_moving)
+
+
+class PVPositionerComparator(PVPositioner):
+    """
+    PV Positioner with a software done signal.
+
+    The done state is set by a comparison function defined in the class body.
+    The comparison function takes two arguments, readback and setpoint,
+    returning True if we are close enough to be considered done or False if we
+    are too far away.
+
+    This class is intended to support `PVPositionerIsClose`, but exists to
+    allow some flexibility if we want to use other metrics for deciding if
+    the PVPositioner is done.
+
+    Internally, this will subscribe to both the `setpoint` and `readback`
+    signals, updating `done` as appropriate.
+    """
+
+    done = Cpt(InternalSignal, value=0)
+    done_value = 1
+
+    def __init__(self, prefix: str, *, name: str, **kwargs):
+        self._last_readback = None
+        self._last_setpoint = None
+        super().__init__(prefix, name=name, **kwargs)
+        if None in (self.setpoint, self.readback):
+            raise NotImplementedError('PVPositionerComparator requires both '
+                                      'a setpoint and a readback signal to '
+                                      'compare!')
+
+    def __init_subclass__(cls, **kwargs):
+        """Set up callbacks in subclass."""
+        super().__init_subclass__(**kwargs)
+        if None not in (cls.setpoint, cls.readback):
+            cls.setpoint.sub_value(cls._update_setpoint)
+            cls.readback.sub_value(cls._update_readback)
+
+    def done_comparator(self, readback: Any, setpoint: Any) -> bool:
+        """
+        Override done_comparator in your subclass.
+
+        This method should return True if we are done moving
+        and False otherwise.
+        """
+        raise NotImplementedError('Must implement a done comparator!')
+
+    def _update_setpoint(self, *args, value: Any, **kwargs) -> None:
+        """Callback to cache the setpoint and update done state."""
+        self._last_setpoint = value
+        # Always set done to False when a move is requested
+        # This means we always get a rising edge when finished moving
+        # Even if the move distance is under our done moving tolerance
+        self.done.put(0, internal=True)
+        self._update_done()
+
+    def _update_readback(self, *args, value: Any, **kwargs) -> None:
+        """Callback to cache the readback and update done state."""
+        self._last_readback = value
+        self._update_done()
+
+    def _update_done(self) -> None:
+        """Update our status to done if we pass the comparator."""
+        if None not in (self._last_readback, self._last_setpoint):
+            is_done = self.done_comparator(self._last_readback,
+                                           self._last_setpoint)
+            self.done.put(int(is_done), internal=True)
```

Comment:
Is it worth only updating if it has changed?  I am 95% sure that the down-stream callbacks are going to fire on over put, I'm not sure ("not done yet", "not done yet", ... )  or ("done", "done", ...) is interesting to most things that would subscribe to the done status?

Reply:
Yes, it's worth modifying this so that it only updates when changed. It's easy to do, makes the situation more clear, and removes burden on downstream callbacks. I'll make this change.